rocAL/rocAL/include/readers/image/external_source_reader.h

```diff
@@ -0,0 +1,123 @@
+    */
+    size_t open() override;
+
+    //! Resets the object's state to read from the first file in the list
+    void reset() override;
+
+    //! Returns the name of the latest file opened
+    std::string id() override { return _last_id; }
+
+    //! Return batch_size() for count_items unless end_of_sequence has been signalled
+    unsigned count_items() override;
+
+    ~ExternalSourceReader() override;
+
+    int close() override;
+    // get_shuffle_time() not applicable for external_source reader
+    unsigned long long get_shuffle_time() { return 0; }
+
+    ExternalSourceReader();
+
+    //! receive next set of filenames from external source
+    void feed_file_names(const std::vector<std::string>& file_names, size_t num_images, bool eos = false) override;
+
+    //! receive next set of file data from external source
+    void feed_data(const std::vector<unsigned char *>& images, const std::vector<size_t>& image_size, ExternalFileMode mode, bool eos = false, int width = 0, int height = 0, int channels = 0) override;
+
+    // mode(): returs the mode for the reader
+    ExternalFileMode mode() { return _file_mode; }
+
+    // get image_dims
+    void get_dims(int cur_idx, int& width, int& height, int& channels);
+
+
+private:
+    //! opens the folder containnig the images
+    std::string _folder_path;
+    std::queue<std::string> _file_names_queue;
+    std::vector<size_t> _file_sizes;
+    std::vector<std::tuple<unsigned char*, size_t, int, int, int>> _file_data;
+    std::queue<std::tuple<unsigned char*, size_t, int, int, int>> _images_data_queue;
+    std::mutex _lock;
+    std::condition_variable _wait_for_input;
+
+    unsigned  _curr_file_idx;
+    FILE* _current_fPtr;
+    unsigned _current_file_size;
+    std::string _last_id;
+    std::string _last_file_name;
+    size_t _shard_id = 0;
+    size_t _shard_count = 1;// equivalent of batch size
+    //!< _batch_count Defines the quantum count of the images to be read. It's usually equal to the user's batch size.
+    /// The loader will repeat images if necessary to be able to have images available in multiples of the load_batch_count,
+    /// for instance if there are 10 images in the dataset and _batch_count is 3, the loader repeats 2 images as if there are 12 images available.
+    size_t _batch_count = 1;
+    size_t _file_id = 0;
+    size_t _in_batch_read_count = 0;
+    bool _loop;
+    bool _shuffle;
+    int _read_counter = 0;
+    volatile bool _end_of_sequence;
+    //!< _file_count_all_shards total_number of files in to figure out the max_batch_size (usually needed for distributed training).
+    void push_file_name(const std::string& image_name);
+    bool pop_file_name(std::string& file_name);
+    void push_file_data(std::tuple<unsigned char*, size_t, int, int, int>& image);
+    bool pop_file_data(std::tuple<unsigned char*, size_t, int, int, int>& image);
+    size_t  _file_count_all_shards;
+    void incremenet_read_ptr();
```

Comment:
spelling check for "increment", below also

Reply:
Resolved